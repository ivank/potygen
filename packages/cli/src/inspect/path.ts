import { Tag, isNode, SqlTag } from '@potygen/ast';
import { PathItem, Path } from '../types';

const isItem =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (item: PathItem<Tag>): item is PathItem<T> =>
    predicate(item.tag);

export const closestParent =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): T | undefined =>
    path.find(isItem(predicate))?.tag;

export const closestParentPath =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): PathItem<T> | undefined =>
    path.find(isItem(predicate));

export const toPath = (tag: Tag, offset: number, path: Path = []): Path | undefined => {
  if (isNode(tag)) {
    const index = (tag.values as Tag[]).findIndex((item) => offset >= item.start && offset <= item.end + 1);

    if (index !== -1) {
      return toPath(tag.values[index], offset, [{ tag, index }, ...path]);
    } else {
      return [{ tag }, ...path];
    }
  }
  return [{ tag }, ...path];
};
