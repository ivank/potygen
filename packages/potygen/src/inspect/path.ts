/**
 * path.ts
 *
 * The ast "path" is a data structure to hold both an sql tag, as well as a "stack" of all of its parents,
 * so we can quickly understand where a tag comes from.
 */

import { findLastIndex } from '../util';
import { isNode } from '../grammar.guards';
import { Tag, SqlTag } from '../grammar.types';
import { PathItem, Path } from '../load.types';

const isItem =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (item: PathItem<Tag>): item is PathItem<T> =>
    predicate(item.tag);

/**
 * Find the closest parent {@link Tag} (using predicate)
 */
export const closestParent =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): T | undefined =>
    path.find(isItem(predicate))?.tag;

/**
 * Find the closest parent {@link Path} (using predicate)
 */
export const closestParentPath =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): PathItem<T> | undefined =>
    path.find(isItem(predicate));

/**
 * Search for the deepest tag within a tree at a given position
 */
export const toPath = (tag: Tag, offset: number, path: Path = []): Path | undefined => {
  if (isNode(tag)) {
    const index = findLastIndex<Tag>((item) => offset >= item.start && offset <= item.end + 1, tag.values);

    if (index !== -1) {
      return toPath(tag.values[index], offset, [{ tag, index }, ...path]);
    } else {
      return [{ tag }, ...path];
    }
  }
  return [{ tag }, ...path];
};
