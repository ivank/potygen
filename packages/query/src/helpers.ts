import { PotygenNotFoundError } from './errors';
import { Sql, SqlMap } from './sql';
import { SqlInterface } from './sql.types';

/**
 * Return the first element, after the query is run, returns undefined if result is empty
 *
 * ```typescript
 * const showSql = sql<ShowSqlQuery>`
 *   SELECT id, name
 *   FROM table1
 *   WHERE id = $id
 * `;
 * const showQuery = maybeOne(showSql);
 *
 * const myRowItem = await showQuery(db, { id: 12 });
 * ```
 */
export const maybeOne = <TQueryInterface extends SqlInterface>(sql: Sql<TQueryInterface>) =>
  new SqlMap<TQueryInterface, TQueryInterface['result'] | undefined>(sql, (rows) =>
    rows.length ? rows[0] : undefined,
  );

/**
 * Return the first element, useful for queries where we always expect at least one result
 *
 * @throws PotygenRuntimeError if row is empty
 *
 * ```typescript
 * const showSql = sql<ShowSqlQuery>`
 *   INSERT (name)
 *   INTO table1
 *   VALUES ($name)
 *   RETURNING id, name
 * `;
 * const showQuery = one(showSql);
 *
 * const myRowItem = await showQuery(db, { name: 'myNewRow' });
 *
 * ```
 */
export const one = <TQueryInterface extends SqlInterface>(sql: Sql<TQueryInterface>) =>
  new SqlMap<TQueryInterface, TQueryInterface['result']>(sql, (rows) => {
    const result = rows[0];
    if (!result) {
      throw new PotygenNotFoundError('Must return at least one', sql.toQueryConfig(rows));
    }
    return result;
  });

/**
 * Return the first element, useful for queries where we always expect at least one result
 *
 * @throws PotygenRuntimeError if row is empty
 *
 * ```typescript
 * const showSql = sql<ShowSqlQuery>`
 *   SELECT id, name
 *   FROM table1
 *   WHERE id = $id
 * `;
 * const idsQuery = map((rows) => rows.map(({ id }) => id)), showSql);
 *
 * const ids = await showQuery(db, { name: 'myNewRow' });
 *
 * ```
 */
export const map = <TQueryInterface extends SqlInterface, TResult>(
  predicate: (rows: TQueryInterface['result'][]) => TResult,
  sql: Sql<TQueryInterface>,
) => new SqlMap<TQueryInterface, TResult>(sql, predicate);
