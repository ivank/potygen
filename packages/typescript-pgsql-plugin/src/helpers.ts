export const correctEmptyIdentifierAfterDot = (sql: string, offset: number): string =>
  sql.charAt(offset - 1) === '.' && sql.charAt(offset).match(/\s/)
    ? sql.slice(0, offset) + 'unknown_column' + sql.slice(offset)
    : sql;
