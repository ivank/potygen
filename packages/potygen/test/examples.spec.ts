import { execSync } from 'child_process';
import { readdirSync, unlinkSync } from 'fs';
import { join } from 'path';
import { connectionString } from './helpers';

const examplesDir = join(__dirname, '../examples/');

const exampleFiles = readdirSync(examplesDir)
  .filter((file) => file.endsWith('.ts'))
  .map((file) => [join('examples', file)]);

describe('Example files', () => {
  beforeAll(() => execSync('yarn tsc', { cwd: examplesDir }));
  afterAll(() =>
    readdirSync(examplesDir)
      .filter((file) => file.endsWith('.js'))
      .forEach((file) => unlinkSync(join(examplesDir, file))),
  );

  it.each(exampleFiles)('Should process %s', (file) => {
    const response = execSync(`yarn node ${file.replace('.ts', '.js')}`, {
      cwd: join(__dirname, '..'),
      env: { ...process.env, POSTGRES_CONNECTION: connectionString },
    });
    expect(response.toString()).toMatchSnapshot(file);
  });
});
