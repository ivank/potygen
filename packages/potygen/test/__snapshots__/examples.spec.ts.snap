// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example files Should process examples/inspect.ts: examples/inspect.ts 1`] = `
"Load all data
Loaded all data
Load views: public.active_reads,public.all_types_view,public.unlevelised_active_reads.
No additional data found, skipping.
Completion Info found in: 
SELECT product FROM orders WHERE region = $region
       ^
       | Column: 7...13

[
  {
    \\"name\\": \\"region\\"
  },
  {
    \\"name\\": \\"product\\"
  },
  {
    \\"name\\": \\"quantity\\"
  },
  {
    \\"name\\": \\"amount\\"
  },
  {
    \\"name\\": \\"orders\\",
    \\"source\\": \\"Table\\"
  }
]
Quick Info found: { type: 'Column', name: 'product', start: 7, end: 13 }:
Quick Info Column: {
  column: { type: 2, postgresType: 'character varying', comment: undefined },
  additionalEnum: undefined,
  additionalComposite: undefined
}:
{
  \\"display\\": \\"product character varying NOT NULL\\",
  \\"description\\": \\"From: orders Table\\",
  \\"start\\": 7,
  \\"end\\": 13
}
"
`;

exports[`Example files Should process examples/load.ts: examples/load.ts 1`] = `
"Load additional data: 1. { tableNames: [ 'public.orders' ] }
Loaded additional data: 1. [ \\"{Table 'public.orders'}\\" ]
{
  \\"params\\": [
    {
      \\"name\\": \\"region\\",
      \\"type\\": {
        \\"type\\": 14,
        \\"nullable\\": true,
        \\"value\\": {
          \\"type\\": 2,
          \\"postgresType\\": \\"character varying\\"
        },
        \\"postgresType\\": \\"any\\"
      }
    }
  ],
  \\"results\\": [
    {
      \\"name\\": \\"product\\",
      \\"type\\": {
        \\"type\\": 2,
        \\"postgresType\\": \\"character varying\\"
      }
    }
  ]
}
"
`;

exports[`Example files Should process examples/load-all.ts: examples/load-all.ts 1`] = `
"Load all data
Loaded all data
Load views: public.active_reads,public.all_types_view,public.unlevelised_active_reads.
No additional data found, skipping.
{
  \\"params\\": [
    {
      \\"name\\": \\"region\\",
      \\"type\\": {
        \\"type\\": 14,
        \\"nullable\\": true,
        \\"value\\": {
          \\"type\\": 2,
          \\"postgresType\\": \\"character varying\\"
        },
        \\"postgresType\\": \\"any\\"
      }
    }
  ],
  \\"results\\": [
    {
      \\"name\\": \\"product\\",
      \\"type\\": {
        \\"type\\": 2,
        \\"postgresType\\": \\"character varying\\"
      }
    }
  ]
}
"
`;

exports[`Example files Should process examples/load-query-interface.ts: examples/load-query-interface.ts 1`] = `
"{
  \\"sources\\": [
    {
      \\"type\\": \\"Table\\",
      \\"isResult\\": true,
      \\"sourceTag\\": {
        \\"tag\\": 90,
        \\"values\\": [
          {
            \\"tag\\": 89,
            \\"values\\": [
              {
                \\"tag\\": 8,
                \\"value\\": \\"users\\",
                \\"start\\": 17,
                \\"end\\": 21
              }
            ],
            \\"start\\": 17,
            \\"end\\": 21
          }
        ],
        \\"start\\": 17,
        \\"end\\": 21
      },
      \\"name\\": \\"users\\",
      \\"table\\": \\"users\\"
    }
  ],
  \\"results\\": [
    {
      \\"name\\": \\"name\\",
      \\"type\\": {
        \\"type\\": 19,
        \\"column\\": \\"name\\",
        \\"sourceTag\\": {
          \\"tag\\": 11,
          \\"values\\": [
            {
              \\"tag\\": 8,
              \\"value\\": \\"name\\",
              \\"start\\": 7,
              \\"end\\": 10
            }
          ],
          \\"start\\": 7,
          \\"end\\": 10
        }
      }
    }
  ],
  \\"params\\": [
    {
      \\"name\\": \\"email\\",
      \\"start\\": 37,
      \\"end\\": 42,
      \\"spread\\": false,
      \\"required\\": false,
      \\"type\\": {
        \\"type\\": 19,
        \\"column\\": \\"email\\",
        \\"sourceTag\\": {
          \\"tag\\": 11,
          \\"values\\": [
            {
              \\"tag\\": 8,
              \\"value\\": \\"email\\",
              \\"start\\": 29,
              \\"end\\": 33
            }
          ],
          \\"start\\": 29,
          \\"end\\": 33
        }
      },
      \\"pick\\": []
    }
  ]
}
"
`;

exports[`Example files Should process examples/parser.ts: examples/parser.ts 1`] = `
"{
  tag: 81,
  values: [
    { tag: 62, values: [Array], start: 7, end: 7 },
    { tag: 69, values: [Array], start: 9, end: 18 }
  ],
  start: 0,
  end: 18
}
"
`;

exports[`Example files Should process examples/sql.ts: examples/sql.ts 1`] = `
"[ { product: 'Sofa' }, { product: 'Chair' }, { product: 'Table' } ]
"
`;

exports[`Example files Should process examples/sql-query.ts: examples/sql-query.ts 1`] = `
"{
  text: 'SELECT product FROM orders WHERE region = $1',
  values: [ 'Sofia' ]
}
"
`;

exports[`Example files Should process examples/static-query-interface.ts: examples/static-query-interface.ts 1`] = `
"[
  {
    name: 'col1',
    type: { type: 3, literal: 123, postgresType: 'float4' }
  }
]
"
`;
